File: mobile/ios/SuperBrainContacts/Sources/API/GraphQLClient.swift
import Foundation

/// Minimal GraphQL client using URLSession for iOS 15+
@available(iOS 15.0, *)
public class GraphQLClient {
    private let baseURL: URL
    private let session: URLSession
    private let timeout: TimeInterval
    
    public init(baseURL: String, timeout: TimeInterval = 30) {
        guard let url = URL(string: baseURL) else {
            fatalError("Invalid base URL: \(baseURL)")
        }
        self.baseURL = url.appendingPathComponent("graphql")
        
        let config = URLSessionConfiguration.default
        config.timeoutIntervalForRequest = timeout
        config.timeoutIntervalForResource = timeout
        self.session = URLSession(configuration: config)
        self.timeout = timeout
    }
    
    /// Execute GraphQL query with optional variables
    /// - Parameters:
    ///   - query: GraphQL query string
    ///   - variables: Optional variables dictionary
    /// - Returns: Raw JSON Data
    /// - Throws: NetworkError or DecodingError
    public func execute(query: String, variables: [String: Any]? = nil) async throws -> Data {
        var request = URLRequest(url: baseURL)
        request.httpMethod = "POST"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        
        var body: [String: Any] = ["query": query]
        if let variables = variables {
            body["variables"] = variables
        }
        
        request.httpBody = try JSONSerialization.data(withJSONObject: body)
        
        let (data, response) = try await session.data(for: request)
        
        guard let httpResponse = response as? HTTPURLResponse else {
            throw NetworkError.invalidResponse
        }
        
        switch httpResponse.statusCode {
        case 200...299:
            return data
        case 400...499:
            throw NetworkError.clientError(statusCode: httpResponse.statusCode, data: data)
        case 500...599:
            throw NetworkError.serverError(statusCode: httpResponse.statusCode)
        default:
            throw NetworkError.unknown(statusCode: httpResponse.statusCode)
        }
    }
}

// MARK: - Error Types

public enum NetworkError: Error, LocalizedError {
    case invalidResponse
    case clientError(statusCode: Int, data: Data)
    case serverError(statusCode: Int)
    case unknown(statusCode: Int)
    
    public var errorDescription: String? {
        switch self {
        case .invalidResponse:
            return "Invalid server response"
        case .clientError(let code, _):
            return "Client error: HTTP \(code)"
        case .serverError(let code):
            return "Server error: HTTP \(code)"
        case .unknown(let code):
            return "Unknown error: HTTP \(code)"
        }
    }
}

File: mobile/android/superbrain-contacts/app/src/main/java/com/superbrain/contacts/api/GraphQLClient.kt
package com.superbrain.contacts.api

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.json.JSONObject
import java.io.IOException
import java.util.concurrent.TimeUnit

/**
 * Minimal GraphQL client using OkHttp with Kotlin coroutines
 */
class GraphQLClient(
    private val baseURL: String,
    timeout: Long = 30
) {
    private val client: OkHttpClient = OkHttpClient.Builder()
        .connectTimeout(timeout, TimeUnit.SECONDS)
        .readTimeout(timeout, TimeUnit.SECONDS)
        .writeTimeout(timeout, TimeUnit.SECONDS)
        .build()
    
    private val graphqlURL = "$baseURL/graphql"
    private val mediaType = "application/json; charset=utf-8".toMediaType()
    
    /**
     * Execute GraphQL query with optional variables
     * @param query GraphQL query string
     * @param variables Optional variables map
     * @return Raw JSON response string
     * @throws NetworkException for network errors
     * @throws IOException for I/O errors
     */
    suspend fun execute(
        query: String,
        variables: Map<String, Any>? = null
    ): String = withContext(Dispatchers.IO) {
        val jsonBody = JSONObject().apply {
            put("query", query)
            variables?.let { put("variables", JSONObject(it)) }
        }
        
        val requestBody = jsonBody.toString().toRequestBody(mediaType)
        
        val request = Request.Builder()
            .url(graphqlURL)
            .post(requestBody)
            .addHeader("Content-Type", "application/json")
            .build()
        
        val response = client.newCall(request).execute()
        
        val responseBody = response.body?.string() 
            ?: throw NetworkException.InvalidResponse("Empty response body")
        
        when (response.code) {
            in 200..299 -> responseBody
            in 400..499 -> throw NetworkException.ClientError(response.code, responseBody)
            in 500..599 -> throw NetworkException.ServerError(response.code)
            else -> throw NetworkException.UnknownError(response.code)
        }
    }
}

/**
 * Network exception types
 */
sealed class NetworkException(message: String) : Exception(message) {
    data class InvalidResponse(val msg: String) : NetworkException("Invalid response: $msg")
    data class ClientError(val statusCode: Int, val data: String) : 
        NetworkException("Client error: HTTP $statusCode")
    data class ServerError(val statusCode: Int) : 
        NetworkException("Server error: HTTP $statusCode")
    data class UnknownError(val statusCode: Int) : 
        NetworkException("Unknown error: HTTP $statusCode")
}
